gateway/broker/session.go

```diff
@@ -0,0 +1,304 @@
+package broker
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/hoophq/hoop/common/log"
+	"github.com/hoophq/hoop/gateway/models"
+)
+
+type Broker struct {
+	agents   sync.Map // map[string]*Connection
+	sessions sync.Map // map[uuid.UUID]*Session
+}
+
+var BrokerInstance = &Broker{}
+
+type Connection struct {
+	ID string `validate:"required,uuid4"`
+	ConnType   string `validate:"required,oneof=websocket tcp"`
+	Connection any
+}
+
+type Session struct {
+	ID uuid.UUID
+	//change this consumer for something else
+	Consumer            *Connection
+	Agent               *Connection
+	Connection          models.Connection
+	clientAddr          string
+	dataChannel         chan []byte
+	credentialsReceived chan bool
+}
+
+func (s *Session) AcknowledgeCredentials() {
+	select {
+	case s.credentialsReceived <- true:
+	default:
+	}
+}
+
+func (s *Session) SendToAgent(data []byte) error {
+	switch s.Agent.ConnType {
+	case "websocket":
+		if wsConn, ok := s.Agent.Connection.(*websocket.Conn); ok {
+			return wsConn.WriteMessage(websocket.BinaryMessage, data)
+		}
+	case "tcp":
+		if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+			_, err := tcpConn.Write(data)
+			return err
+		}
+	default:
+		return nil
+	}
+	return nil
+}
+
+func (s *Session) ReadFromAgent() (int, []byte, error) {
+	switch s.Agent.ConnType {
+	case "websocket":
+		if wsConn, ok := s.Agent.Connection.(*websocket.Conn); ok {
+			_, message, err := wsConn.ReadMessage()
+			return len(message), message, err
+		}
+	case "tcp":
+		if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+			buffer := make([]byte, 16*1024)
+			n, err := tcpConn.Read(buffer)
+			if err != nil {
+				return 0, nil, err
+			}
+			return n, buffer, nil
+		}
+	default:
+		return 0, nil, nil
+	}
+	return 0, nil, nil
+}
+
+func (s *Session) Close() {
+	close(s.dataChannel)
+	if s.Consumer != nil {
+		if conn, ok := s.Consumer.Connection.(net.Conn); ok {
+			conn.Close()
+		}
+	}
+	if s.Agent != nil {
+		switch s.Agent.ConnType {
+		case "websocket":
+			if wsConn, ok := s.Agent.Connection.(*websocket.Conn); ok {
+				wsConn.Close()
+			}
+		case "tcp":
+			if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+				tcpConn.Close()
+			}
+		}
+	}
+	BrokerInstance.sessions.Delete(s.ID)
+	log.Infof("Session %s closed", s.ID)
+}
+
+// forward data from agent to tcp
+func (s *Session) ForwardToTCP(data []byte) {
+	log.Infof("Forwarded %d bytes to TCP session %s", len(data), s.ID)
+	s.dataChannel <- data
+}
+
+// this will spam data from tcp to agent wsconn
+func (s *Session) StartingForwardind(data []byte) error {
```

Comment:
this whole session file is couple on tcp/ws/rdp protocol now we can think in a way to abstract like session has the type and its type has its one instance that handle the protocol internals